spark/src/main/scala/org/apache/spark/sql/delta/TableFeature.scala

```diff
@@ -284,28 +282,37 @@ object TableFeature {
     allSupportedFeaturesMap.get(featureName.toLowerCase(Locale.ROOT))
 
   /**
-   * Extracts the downgraded feature name from protocol downgrade commit info.
+   * Extracts the removed feature name by comparing current and base protocols.
+   * Returns None if there is no removed feature.
    */
-  def getRemovedFeatureName(commitInfoOpt: Option[CommitInfo]): Option[String] = {
-    commitInfoOpt
-      .flatMap(_.tags)
-      .flatMap(_.get(TableFeature.DROP_FEATURE_COMMIT_INFO_TAG))
+  protected def getRemovedFeatureName(
+      currentProtocol: Protocol,
+      baseProtocol: Protocol): Option[String] = {
+    val currentFeatureNames = currentProtocol.readerAndWriterFeatureNames
+    val baseFeatureNames = baseProtocol.readerAndWriterFeatureNames
+    val removedFeatures = baseFeatureNames -- currentFeatureNames
+
+    require(removedFeatures.size <= 1)
+    Option(baseFeatureNames -- currentFeatureNames).filter(_.nonEmpty).map(_.head)
   }
 
   /**
-   * Identifies from CommitInfo whether this is a feature removal commit.
+   * Identifies whether there was a feature removal between two protocols.
    */
-  def isDropFeatureCommit(commitInfoOpt: Option[CommitInfo]): Boolean =
-    getRemovedFeatureName(commitInfoOpt).isDefined
+  def isDropFeatureCommit(currentProtocol: Protocol, baseProtocol: Protocol): Boolean = {
+    getRemovedFeatureName(currentProtocol, baseProtocol).isDefined
```

Comment:
Same here -- let's make this work for multiple features.

Furthermore -- what should happen if we are comparing across protocol versions? To create a generic function, we could just get the implicitly supported feature names, and compare those. In that case the caller should still explicitly guard against protocol version reduction, which I think it does. The advantage is that this function would then be _actually generic_ and do what it actually says that it does, instead of having a generic-sounding name but actually having a much more specific and narrow use case and implementation.

Reply:
> Furthermore -- what should happen if we are comparing across protocol versions? To create a generic function, we could just get the implicitly supported feature names, and compare those. In that case the caller should still explicitly guard against protocol version reduction, which I think it does. The advantage is that this function would then be _actually generic_ and do what it actually says that it does, instead of having a generic-sounding name but actually having a much more specific and narrow use case and implementation.

The trouble here is that implicit features are only populated when the protocol does not support table features. So if someone attempts to downgrade a protocol that supports table feature to one that does not the diff won't work. Instead of complicating the function to cover all the cases I think is better to narrow the scope, e.g. `getExplicitDroppedFeatures`.